packages/flash/src/index.test.ts

```diff
@@ -0,0 +1,95 @@
+import { assert } from 'chai';
+
+import { flashMiddleware, flash } from './index';
+
+describe('flash', () => {
+  it('throws an error if no session present', () => {
+    assert.throw(() => {
+      flashMiddleware()({} as any, {} as any, () => {});
+    });
+  });
+
+  it('throws an error when middleware is not used', () => {
+    assert.throw(() => {
+      flash('notice', 'Hello world');
+    });
+  });
+
+  it('adds a flash', () => {
+    const req = {
+      session: {},
+    } as any;
+    const res = {} as any;
+
+    flashMiddleware()(req, res, () => {
+      flash('notice', 'hello world');
+
+      assert.sameDeepMembers(flash(), [{ type: 'notice', message: 'hello world' }]);
+    });
+  });
+
+  it('overwrites flash with the same type', () => {
+    const req = {
+      session: {},
+    } as any;
+    const res = {} as any;
+
+    flashMiddleware()(req, res, () => {
+      flash('notice', 'hello world');
+      flash('notice', 'goodbye world');
+
+      assert.sameDeepMembers(flash(), [{ type: 'notice', message: 'goodbye world' }]);
+    });
+  });
+
+  it('returns flash message for a given type', () => {
+    const req = {
+      session: {},
+    } as any;
+    const res = {} as any;
+
+    flashMiddleware()(req, res, () => {
+      flash('notice', 'hello world');
+      flash('error', 'goodbye world');
+
+      assert.equal(flash('notice'), 'hello world');
+      assert.equal(flash('error'), 'goodbye world');
+    });
+  });
+
+  it('returns all flashes', () => {
+    const req = {
+      session: {},
+    } as any;
+    const res = {} as any;
+
+    flashMiddleware()(req, res, () => {
+      flash('notice', 'hello world');
+      flash('error', 'goodbye world');
+
+      assert.sameDeepMembers(flash(), [
+        { type: 'notice', message: 'hello world' },
+        { type: 'error', message: 'goodbye world' },
+      ]);
+    });
+  });
+
+  it('clears flash after it has been read', () => {
```

Comment:
I should probably document this (I need to add a README to this package), but this is subtly different from what other frameworks (Rails/Laravel/etc.) do. Other frameworks typically persist flash messages *only* until the very next HTTP request, at which point they're cleared, whether or not they've been read. On the other hand, this implementation persists the messages until they're actually read. This is probably fine in practice?

For context, the API I used here is modeled after https://www.npmjs.com/package/connect-flash. I didn't use that package directly because I wanted to use `AsyncLocalStorage` so we don't have to pass around a reference to `req.session` everywhere.

Reply:
> I should probably document this (I need to add a README to this package), but this is subtly different from what other frameworks (Rails/Laravel/etc.) do. Other frameworks typically persist flash messages _only_ until the very next HTTP request, at which point they're cleared, whether or not they've been read. On the other hand, this implementation persists the messages until they're actually read. This is probably fine in practice?

I actually prefer the option used here. Other HTTP requests may, due to weird timing, end up being an unrelated request (e.g., favicon, deferred/delayed loading of a script/image/style of the previous page, AJAX request from a different tab, etc.).

Reply:
I agree that this seems like a better system.